source/agora/node/TransactionRelayer.d

```diff
@@ -0,0 +1,462 @@
+        public TimePoint expiry;
+
+        ///
+        public static bool cmpFeeLess (in TxHolder tx_holder1, in TxHolder tx_holder2) @safe nothrow
+        {
+            return tx_holder1.fee > tx_holder2.fee;
+        }
+
+        ///
+        public void toString (scope SinkT dg) const @trusted
+        {
+            import std.format;
+
+            // Workaround for missing @safe on the delegate passed to RedBlackTree.toString().
+            auto safe_dg = cast(SafeSinkT) dg;
+
+            formattedWrite(safe_dg, "{ amount: ");
+            this.fee.toString(safe_dg);
+            formattedWrite(safe_dg, ", hash: ");
+            this.hash.toString(safe_dg);
+            formattedWrite(safe_dg, ", expiry: %d }", this.expiry);
+        }
+
+        ///
+        public string toString () const @safe
+        {
+            string ret;
+            void delegate(scope const(char)[] v) mydg ;
+            scope SinkT dg = mydg;
+            this.toString(dg);
+            return ret;
+        }
+    }
+
+    /***************************************************************************
+
+        Constructor used in tests
+
+        Params:
+            config = config
+
+    ***************************************************************************/
+
+    version (unittest)
+    private this (immutable Config config)
+    {
+        import agora.common.ManagedDatabase;
+        import agora.consensus.Fee;
+        import agora.consensus.data.Params;
+
+        auto stateDB = new ManagedDatabase(":memory:");
+        auto cacheDB = new ManagedDatabase(":memory:");
+        immutable params = new immutable(ConsensusParams)();
+
+        this.config = config;
+        this.clock = new MockClock(0);
+        this.pool = new TransactionPool(cacheDB);
+        this.utxo_set = new TestUTXOSet();
+        auto fee_man = new FeeManager(stateDB, params);
+        this.getAdjustedTXFee = (in Transaction tx, out Amount tot_fee)
+        {
+            return fee_man.getAdjustedTXFee(tx, &utxo_set.peekUTXO, tot_fee);
+        };
+
+        this.log = Logger(__MODULE__);
+        this.txs = new RedBlackTree!(TxHolder, TxHolder.cmpFeeLess, true)();
+    }
+
+    /***************************************************************************
+
+        Utility method to create a transaction with a fee.
+
+        Params:
+            gen_out_ind = output index of the first Genesis payment transaction
+            fee = the required fee in the transaction
+
+        Returns:
+            transaction with the desired fee
+
+    ***************************************************************************/
+
+    version (unittest)
+    static Transaction getTX (ushort gen_out_ind, Amount fee)
+    {
+        import agora.consensus.data.genesis.Test;
+        import agora.crypto.Key;
+        import std.array;
+
+        auto genesis_tx = GenesisBlock.txs.filter!(tx => tx.type == TxType.Payment).array()[0];
```

Comment:
Instead of `array()[0]` which does an allocation just to take the first element, use `front`.